docs/conceptual/core/executor.md

```diff
@@ -2,8 +2,147 @@
 title: Executor
 ---
 
-* Runtime Versioning
+The executor is responsible for dispatching and executing calls into the
+Substrate runtime.
 
-* Native VS Wasm Execution
+## Runtime Execution
 
-* Forkless Runtime Upgrades
\ No newline at end of file
+The Substrate runtime is compiled into a native executable and a WebAssembly
+(Wasm) binary.
+
+The native runtime is included as part of the node executable, while the Wasm
+binary is stored on the blockchain under a well known storage key.
+
+These two representations of the runtime may not be the same. For example: when
+the runtime is upgraded. The executor determines which version of the runtime to
+use when dispatching and executing calls.
+
+### Execution Strategy
+
+Before runtime execution begins, the Substrate client proposes which runtime
+execution environment should be used. This is controlled by the execution
+strategy, which can be configured for the different parts of the blockchain
+execution process. The strategies are:
+
+- `Native`: Execute with native build (if available, WebAssembly otherwise).
+- `Wasm`: Only execute with the WebAssembly build.
+- `Both`: Execute with both native (where available) and WebAssembly builds.
+- `NativeElseWasm`: Execute with the native build if possible; if it fails, then
+  execute with WebAssembly.
+
+The default execution strategies for the different parts of the blockchain
+execution process are:
+
+- Syncing: `NativeElseWasm`
+- Block Import: `NativeElseWasm`
+- Block Construction: `Wasm`
+- Off-Chain Worker: `Native`
+- Other: `Native`
+
+### Wasm Execution
+
+The Wasm representation of the Substrate runtime is considered the canonical
+runtime. Because this Wasm runtime is placed in the blockchain storage, the
+network must come to consensus about this binary. Thus it can be verified to be
+consistent across all syncing nodes.
+
+The Wasm execution environment can be more restrictive than the native execution
+environment. For example, the Wasm runtime always executes in a 32-bit
+environment with a 4 GB memory limit.
```

Comment:
The 4 GB memory limit is useful information to have on other docs.

Reply:
What other docs?

Reply:
Maybe the high level docs or the recipes. I'm just saying this because on one of my modules I have a Vec as a storage value. That Vec could reach 4 GB, knowing that there is a 4 GB limit I can try another solution. Or maybe adding checks to make sure that the Vec doesn't reach 4 GB.

Reply:
Now that I think about, having the information that it is a 32-bit environment is also important. A Vec can easily reach 2**32 elements. This would make the runtime panic. So, I would also need to add a check to prevent the Vec from reaching 2**32 elements.

Reply:
How can a vec easily reach 2^32 elements? You will run into problems much earlier. Such a huge state, that need to be synchronized between all nodes, will kill your blockchain already much earlier.

Reply:
Hmm, I will correct my previous position. If a vec reaches 2^32 elements, given that the minimum element size is 1 byte, it would panic because it would exceed the 4 GB limit.
But other than that, reaching 2^32 elements or 4 GB is not that hard in a blockchain. 2^32 is ~4 billion. Ethereum currently has 80 million addresses. That's only a difference of 50x, it's not hard to imagine that it will reach 4 billion over the next few years.
Also, I couldn't find the current size of Ethereum's state (and so correct me if I'm wrong), but with 80 million addresses it has to be bigger than 4 GB by now (an average size of 50 bytes per address would be enough to exceed 4 GB).

Reply:
No one says that the state can not be bigger, you can just not load 80 million addresses at once. That would not work from the memory perspective and also you would not have the computational power to load all these addresses in the small time frame you have for building a block.

Reply:
yeah -- even with an infinitely large state, if there is a code that wants to load 4GB into a Vec, and that code fails, I don't really blame the system but rather the code. In the context of a blockchain that really should not happen. 

Maybe offchain which also runs in wasm could be a catch here.

Reply:
Hmm, ok. I think I misunderstood the memory limit. Then it's a limit on the working memory, not on the state size. If instead of using a Vec, I implement the same data structure using a StorageMap and a StorageValue, I don't have any limit on size or number of elements. Am I thinking correctly?

Reply:
Yes.

Reply:
@brunoffranca I think the thing you are asking for, and makes sense, is a separate page just on Wasm and the Wasm environment in Substrate.

Maybe the set of assumptions about Wasm in Substrate, limitations/settings, difference between Contract and Runtime and Offchain Worker, etc... Use of Wasmi and soon Lightbeam.

Maybe there is even a seperate Wasm section which has all these subsections.

Would that solve your ask?

Reply:
Well, my initial concern about the Wasm execution environment was unfounded. And even if it wasn't, the limits would only matter in edge cases (ie: my whole thing with the 4 GB Vec). So an entire page on the Wasm environment will probably not be useful for most people (who just want to learn how to create their own modules).
That being said, it would be something interesting to read. I always like to learn what I can about the Substrate core, even if I don't really interact with it.